### PR TITLE
Fix some possible thread-id overflow calculations

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes some possible thread-id calculations or usages that may possibly overflow `int32` type or `size_type`.
Reference #10368 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/partitioning/partitioning.cu [hunk 1]
```diff
@@ -138,7 +138,7 @@ CUDF_KERNEL void compute_row_partition_numbers(row_hasher_t the_hasher,
   auto const stride = cudf::detail::grid_1d::grid_stride();
 
   // Initialize local histogram
-  size_type partition_number = threadIdx.x;
+  thread_index_type partition_number = threadIdx.x;
   while (partition_number < num_partitions) {
     shared_partition_sizes[partition_number] = 0;
     partition_number += blockDim.x;
```
Comment (davidwendt) [inline cpp/src/partitioning/partitioning.cu:144]: If `num_partitions` is close to `max<size_type>`  then `partition_number += blockDim.x` could overlfow `size_type`.

**Review by vuule — APPROVED** [review]

**Review by shrshi — APPROVED** [review]
Looks good to me!

**Pull request merged**